File: frontend/src/lib/lingua.ts
/* =============================================================================
 * Montaigne · risoluzione della lingua dell'interfaccia
 *
 * PRD: "interfaccia bilingue italiano e inglese dal primo giorno... date e
 * numeri seguono la lingua del browser" — nessun selettore (design-frontend.md
 * non ne descrive uno, §23 rimanda solo il perimetro della traduzione, non il
 * meccanismo). La lingua è quindi una funzione pura della stessa intestazione
 * `Accept-Language` che il browser manda da solo con ogni richiesta, calcolata
 * lato server a ogni richiesta — stesso stile di `lib/light.ts`, che calcola
 * la luce dall'ora invece di leggerla da un timer client: nessun cookie,
 * nessun `localStorage`, nessuno stato da tenere sincronizzato.
 *
 * Il backend risolve la stessa intestazione con lo stesso algoritmo
 * (`backend/app/core/lingua.py`): è ciò che tiene allineata la lingua in cui
 * next-intl mostra l'interfaccia con la lingua in cui il backend sceglie fra
 * le varianti di titolo/descrizione/etichetta di genere già salvate nelle due
 * lingue (design-frontend.md, entità Variante di titolo/Descrizione) — issue
 * #34, "Attenzione": una sola preferenza, non due che possono disallinearsi.
 * ========================================================================== */

export const LINGUE_INTERFACCIA = ["it", "en"] as const;

export type LinguaInterfaccia = (typeof LINGUE_INTERFACCIA)[number];

export const LINGUA_PREDEFINITA: LinguaInterfaccia = "it";

function èLinguaInterfaccia(codice: string): codice is LinguaInterfaccia {
  return (LINGUE_INTERFACCIA as readonly string[]).includes(codice);
}

/**
 * La prima lingua fra quelle richieste (in ordine di preferenza, `;q=`
 * incluso) che compare fra le due supportate; altrimenti l'italiano. Stesso
 * algoritmo di `backend/app/core/lingua.py::_lingua_da_intestazione` — le due
 * implementazioni vanno tenute allineate a mano, non condividono codice
 * perché vivono in runtime diversi.
 */
export function risolviLingua(acceptLanguage: string | null | undefined): LinguaInterfaccia {
  if (!acceptLanguage) return LINGUA_PREDEFINITA;

  const voci = acceptLanguage
    .split(",")
    .map((parte) => {
      const [tag, qParam] = parte.trim().split(";q=");
      const codice = (tag ?? "").trim().split("-")[0]?.toLowerCase() ?? "";
      // `qParam === undefined` e non un controllo di verità: `;q=` presente
      // ma vuoto (`"it;q=,en"`, di fatto malformato) deve valere qualità 0,
      // non 1 — altrimenti una stringa vuota, falsy quanto "assente", si
      // confonde con l'assenza di `;q=` e vince quando non dovrebbe. Stesso
      // esito di `backend/app/core/lingua.py`, dove `float("")` solleva e
      // ricade sullo stesso 0 — le due implementazioni vanno tenute
      // allineate anche su questo caso limite, non solo sul percorso comune.
      const qualita = qParam === undefined ? 1 : Number(qParam);
      return { codice, qualita: Number.isFinite(qualita) ? qualita : 0 };
    })
    .sort((a, b) => b.qualita - a.qualita);

  for (const { codice } of voci) {
    if (èLinguaInterfaccia(codice)) return codice;
  }
  return LINGUA_PREDEFINITA;
}

/**
 * Le intestazioni comuni di ogni fetcher di `lib/api/*.ts` verso il backend:
 * l'autenticazione, più `Accept-Language` quando il chiamante la passa
 * esplicitamente (i fetcher usati da un Server Component per il rendering
 * iniziale — vedi `lib/api/lingua-richiesta.ts`). Un fetch dal browser non
 * passa nulla qui: la manda già da sé, non richiede questo aiuto.
 */
export function intestazioniConLingua(
  accessToken: string,
  acceptLanguage?: string,
): HeadersInit {
  return {
    Authorization: `Bearer ${accessToken}`,
    ...(acceptLanguage ? { "Accept-Language": acceptLanguage } : {}),
  };
}

File: backend/app/core/lingua.py
"""Risoluzione della lingua dell'interfaccia lato server (issue #34), dalla
stessa `Accept-Language` che il browser manda da solo con ogni richiesta —
nessuna preferenza indipendente da quella che il frontend applica con
next-intl (`frontend/src/lib/lingua.ts`, stesso algoritmo in TypeScript, da
tenere allineato a mano: le due implementazioni vivono in runtime diversi e
non condividono codice).

Sostituisce le costanti `_LINGUA_INTERFACCIA = "it"` finora duplicate in
`voce_repository.py`, `metriche_service.py` ed `export_service.py`: la
scelta fra le varianti di titolo/descrizione/etichetta di genere già salvate
nelle due lingue diventa un parametro reale, iniettato dai router con
`Depends(lingua_interfaccia)` e propagato esplicitamente fino al punto in
cui si sceglie la riga — mai un valore fisso.

`app/repositories/preview_repository.py` resta fuori: la sua
`LINGUA_INTERFACCIA` alimenta il contesto mandato al modello per le funzioni
assistite (preview personalizzata, sintesi, suggerimenti), non un testo
mostrato direttamente nell'interfaccia — un problema distinto, non ancora
nel perimetro di questa issue.
"""

from fastapi import Header

LINGUE_INTERFACCIA = ("it", "en")
LINGUA_PREDEFINITA = "it"


def _lingua_da_intestazione(valore: str | None) -> str:
    """La prima lingua fra quelle richieste (in ordine di preferenza,
    qualità `;q=` inclusa) che compare fra le due supportate; altrimenti
    l'italiano — stesso comportamento della costante che sostituisce."""
    if not valore:
        return LINGUA_PREDEFINITA

    voci: list[tuple[float, str]] = []
    for parte in valore.split(","):
        pezzi = parte.strip().split(";q=")
        codice = pezzi[0].strip().split("-")[0].lower()
        try:
            qualita = float(pezzi[1]) if len(pezzi) > 1 else 1.0
        except ValueError:
            qualita = 0.0
        voci.append((qualita, codice))

    for _, codice in sorted(voci, key=lambda voce: voce[0], reverse=True):
        if codice in LINGUE_INTERFACCIA:
            return codice
    return LINGUA_PREDEFINITA


async def lingua_interfaccia(accept_language: str | None = Header(default=None)) -> str:
    """Dipendenza FastAPI: `Depends(lingua_interfaccia)` in un router."""
    return _lingua_da_intestazione(accept_language)
